src/canonical.rs

```diff
@@ -0,0 +1,421 @@
+        assert_eq!(canon("AGC"), "ATC");
+        assert_eq!(canon("AGG"), "AAT");
+
+        assert_eq!(canon("TAA"), "AAT");
+        assert_eq!(canon("TAT"), "ATA");
+        assert_eq!(canon("TAC"), "ATC");
+        assert_eq!(canon("TAG"), "ATC");
+        assert_eq!(canon("TTA"), "AAT");
+        assert_eq!(canon("TTT"), "AAA");
+        assert_eq!(canon("TTC"), "AAT");
+        assert_eq!(canon("TTG"), "AAT");
+        assert_eq!(canon("TCA"), "ATC");
+        assert_eq!(canon("TCT"), "ATA");
+        assert_eq!(canon("TCC"), "AAT");
+        assert_eq!(canon("TCG"), "ATC");
+        assert_eq!(canon("TGA"), "ATC");
+        assert_eq!(canon("TGT"), "ATA");
+        assert_eq!(canon("TGC"), "ATC");
+        assert_eq!(canon("TGG"), "AAT");
+
+        assert_eq!(canon("CAA"), "AAT");
+        assert_eq!(canon("CAT"), "ATC");
+        assert_eq!(canon("CAC"), "ATA");
+        assert_eq!(canon("CAG"), "ATC");
+        assert_eq!(canon("CTA"), "ATC");
+        assert_eq!(canon("CTT"), "AAT");
+        assert_eq!(canon("CTC"), "ATA");
+        assert_eq!(canon("CTG"), "ATC");
+        assert_eq!(canon("CCA"), "AAT");
+        assert_eq!(canon("CCT"), "AAT");
+        assert_eq!(canon("CCC"), "AAA");
+        assert_eq!(canon("CCG"), "AAT");
+        assert_eq!(canon("CGA"), "ATC");
+        assert_eq!(canon("CGT"), "ATC");
+        assert_eq!(canon("CGC"), "ATA");
+        assert_eq!(canon("CGG"), "AAT");
+
+        assert_eq!(canon("GAA"), "AAT");
+        assert_eq!(canon("GAT"), "ATC");
+        assert_eq!(canon("GAC"), "ATC");
+        assert_eq!(canon("GAG"), "ATA");
+        assert_eq!(canon("GTA"), "ATC");
+        assert_eq!(canon("GTT"), "AAT");
+        assert_eq!(canon("GTC"), "ATC");
+        assert_eq!(canon("GTG"), "ATA");
+        assert_eq!(canon("GCA"), "ATC");
+        assert_eq!(canon("GCT"), "ATC");
+        assert_eq!(canon("GCC"), "AAT");
+        assert_eq!(canon("GCG"), "ATA");
+        assert_eq!(canon("GGA"), "AAT");
+        assert_eq!(canon("GGT"), "AAT");
+        assert_eq!(canon("GGC"), "AAT");
+        assert_eq!(canon("GGG"), "AAA");
+    }
+
+    #[test]
+    fn canonicalization_selects_reverse_if_that_is_lexically_less() {
+        assert_eq!(canon("TTGT"), "AATA");
+        assert_eq!(canon("TGTT"), "AATA");
+
+        assert_eq!(canon("ATCGCCAT"), "ATCCGCAT");
+    }
+
+    quickcheck! {
+        fn forward_canonicalization_is_idempotent(dna: DnaSequenceStrict) -> bool {
+            let canonical = ForwardCanonical::new(dna.as_slice().iter().copied());
+            let canonical2 = ForwardCanonical::new(canonical.clone());
+            canonical2.eq(canonical)
+        }
+
+        fn canonicalization_is_idempotent(dna: DnaSequenceStrict) -> bool {
+            // Need full allocation because canonical needs a double-ended iter, but isn't one itself.
+            let canonical = dna.canonical();
+            let canonical2 = canonical.canonical();
+            canonical2 == canonical
+        }
+    }
```

Comment:
Do we want to quickcheck `LexicalMin` as well? Could compare it to the allocation version with `Vec` as a sanity check.